substrate/frame/identity/src/migration.rs

```diff
@@ -0,0 +1,124 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Storage migrations for the Identity pallet.
+
+use super::*;
+use frame_support::{migrations::VersionedMigration, pallet_prelude::*, traits::OnRuntimeUpgrade};
+
+#[cfg(feature = "try-runtime")]
+use codec::{Decode, Encode};
+#[cfg(feature = "try-runtime")]
+use sp_runtime::TryRuntimeError;
+
+pub mod versioned {
+	use super::*;
+
+	pub type V0ToV1<T, const KL: u64> = VersionedMigration<
+		0,
+		1,
+		v1::VersionUncheckedMigrateV0ToV1<T, KL>,
+		crate::pallet::Pallet<T>,
+		<T as frame_system::Config>::DbWeight,
+	>;
+}
+
+pub mod v1 {
+	use super::*;
+
+	/// The log target.
+	const TARGET: &'static str = "runtime::identity::migration::v1";
+
+	/// The old identity type, useful in pre-upgrade.
+	mod v0 {
+		use super::*;
+		use frame_support::storage_alias;
+
+		#[storage_alias]
+		pub type IdentityOf<T: Config> = StorageMap<
+			Pallet<T>,
+			Twox64Concat,
+			<T as frame_system::Config>::AccountId,
+			Registration<
+				BalanceOf<T>,
+				<T as pallet::Config>::MaxRegistrars,
+				<T as pallet::Config>::IdentityInformation,
+			>,
+			OptionQuery,
+		>;
+	}
+
+	/// Migration to add usernames to Identity info.
+	///
+	/// `T` is the runtime and `KL` is the key limit to migrate. This is just a safety guard to
+	/// prevent stalling a parachain by accumulating too much weight in the migration. To have an
+	/// unlimited migration (e.g. in a chain without PoV limits), set this to `u64::MAX`.
+	pub struct VersionUncheckedMigrateV0ToV1<T, const KL: u64>(PhantomData<T>);
+	impl<T: Config, const KL: u64> OnRuntimeUpgrade for VersionUncheckedMigrateV0ToV1<T, KL> {
+		#[cfg(feature = "try-runtime")]
+		fn pre_upgrade() -> Result<Vec<u8>, TryRuntimeError> {
+			let identities = v0::IdentityOf::<T>::iter().count();
+			log::info!(
+				target: TARGET,
+				"pre-upgrade state contains '{}' identities.",
+				identities
+			);
+			ensure!((identities as u64) < KL, "too many identities to migrate");
+			Ok((identities as u64).encode())
+		}
+
+		fn on_runtime_upgrade() -> Weight {
+			log::info!(
+				target: TARGET,
+				"running storage migration from version 0 to version 1."
+			);
+
+			let mut weight = T::DbWeight::get().reads(1);
+			let mut translated: u64 = 0;
+			let mut interrupted = false;
+
+			for (account, registration) in v0::IdentityOf::<T>::iter() {
+				IdentityOf::<T>::insert(account, (registration, None::<Username<T>>));
+				translated.saturating_inc();
+				if translated >= KL {
+					log::warn!(
+						"Incomplete! Migration limit reached. Only {} identities migrated.",
+						translated
+					);
+					interrupted = true;
+					break
+				}
+			}
+			if !interrupted {
+				log::info!("all {} identities migrated", translated);
+			}
+
+			weight.saturating_accrue(T::DbWeight::get().reads_writes(translated, translated));
+			weight.saturating_accrue(T::DbWeight::get().writes(1));
+			weight
+		}
```

Comment:
This migration is missing a storage version bump from v0 to v1. How can this slip through if even `try-runtime` prints an error after applying the migration?

```
[2024-05-29T08:46:53Z INFO  runtime::identity::migration::v1] pre-upgrade state contains '35' identities.
[2024-05-29T08:46:53Z INFO  runtime::identity::migration::v1] running storage migration from version 0 to version 1.
[2024-05-29T08:46:53Z INFO  pallet_identity::migration::v1] all 35 identities migrated
[2024-05-29T08:46:53Z INFO  pallet_identity::migration::v1] post-upgrade expects '35' identities to have been migrated.
[2024-05-29T08:46:53Z INFO  runtime::identity::migration::v1] migrated all identities.
[2024-05-29T08:46:53Z ERROR runtime::frame-support] Identity: On chain storage version StorageVersion(0) doesn't match current storage version StorageVersion(1).
```

I hope this is an exception because manually bumping storage versions should be avoided at all costs.

Reply:
Update: Sorry, I see this is covered by the `versioned` module wrapper.